File: conduit/tests/python/test_arrays_choices.py
"""Tests for array and choice type encode/decode in the Conduit generated Python codecs."""
import struct
import pytest

from arrays_choices import (
    BitReader,
    BitWriter,
    DecodeError,
    Constants,
)
from arrays_choices.structs import Point, SubX, SubY, TypeABody, TypeBBody, FallbackBody
from arrays_choices.messages import FixedArrayMsg, CountFromArrayMsg, ChoiceMsg


# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------
class TestArraysChoicesConstants:

    def test_type_a_value(self):
        assert Constants.TYPE_A == 1

    def test_type_b_value(self):
        assert Constants.TYPE_B == 2

    def test_sub_x_value(self):
        assert Constants.SUB_X == 10

    def test_sub_y_value(self):
        assert Constants.SUB_Y == 20


# ---------------------------------------------------------------------------
# Point struct
# ---------------------------------------------------------------------------
class TestPointStruct:

    def test_default_values(self):
        p = Point()
        assert p.x == 0
        assert p.y == 0

    def test_roundtrip(self):
        p = Point()
        p.x = 100
        p.y = 200
        data = p.encode_bytes()
        p2 = Point.decode_bytes(data)
        assert p2.x == 100
        assert p2.y == 200

    def test_max_values(self):
        p = Point()
        p.x = 0xFFFF
        p.y = 0xFFFF
        data = p.encode_bytes()
        p2 = Point.decode_bytes(data)
        assert p2.x == 0xFFFF
        assert p2.y == 0xFFFF

    def test_zero_values(self):
        p = Point()
        p.x = 0
        p.y = 0
        data = p.encode_bytes()
        p2 = Point.decode_bytes(data)
        assert p2.x == 0
        assert p2.y == 0

    def test_wire_size(self):
        """Each Point has two u16 fields = 4 bytes."""
        p = Point()
        p.x = 1
        p.y = 2
        data = p.encode_bytes()
        assert len(data) == 4

    def test_encode_big_endian(self):
        p = Point()
        p.x = 0x1234
        p.y = 0x5678
        data = p.encode_bytes()
        assert data == bytes([0x12, 0x34, 0x56, 0x78])

    def test_repr(self):
        p = Point()
        p.x = 10
        p.y = 20
        assert "10" in repr(p)
        assert "20" in repr(p)


# ---------------------------------------------------------------------------
# FixedArrayMsg
# ---------------------------------------------------------------------------
class TestFixedArrayMsg:

    def test_roundtrip_with_three_points(self):
        msg = FixedArrayMsg()
        for i in range(3):
            p = Point()
            p.x = (i + 1) * 10
            p.y = (i + 1) * 20
            msg.points.append(p)

        data = msg.encode_bytes()
        msg2 = FixedArrayMsg.decode_bytes(data)
        assert len(msg2.points) == 3
        assert msg2.points[0].x == 10
        assert msg2.points[0].y == 20
        assert msg2.points[1].x == 20
        assert msg2.points[1].y == 40
        assert msg2.points[2].x == 30
        assert msg2.points[2].y == 60

    def test_wire_size(self):
        """3 Points * 4 bytes each = 12 bytes."""
        msg = FixedArrayMsg()
        for _ in range(3):
            msg.points.append(Point())
        data = msg.encode_bytes()
        assert len(data) == 12

    def test_decode_always_reads_three(self):
        """Regardless of what we put in, decode reads exactly 3 Points."""
        data = bytes(12)  # 3 Points of zeros
        msg = FixedArrayMsg.decode_bytes(data)
        assert len(msg.points) == 3

    def test_idempotent(self):
        msg = FixedArrayMsg()
        for i in range(3):
            p = Point()
            p.x = i * 100
            p.y = i * 200
            msg.points.append(p)
        data1 = msg.encode_bytes()
        msg2 = FixedArrayMsg.decode_bytes(data1)
        data2 = msg2.encode_bytes()
        assert data1 == data2

    def test_max_value_points(self):
        msg = FixedArrayMsg()
        for _ in range(3):
            p = Point()
            p.x = 0xFFFF
            p.y = 0xFFFF
            msg.points.append(p)
        data = msg.encode_bytes()
        msg2 = FixedArrayMsg.decode_bytes(data)
        for pt in msg2.points:
            assert pt.x == 0xFFFF
            assert pt.y == 0xFFFF

    def test_zero_array(self):
        """All zeros should decode fine."""
        msg = FixedArrayMsg()
        for _ in range(3):
            msg.points.append(Point())
        data = msg.encode_bytes()
        msg2 = FixedArrayMsg.decode_bytes(data)
        for pt in msg2.points:
            assert pt.x == 0
            assert pt.y == 0


# ---------------------------------------------------------------------------
# CountFromArrayMsg
# ---------------------------------------------------------------------------
class TestCountFromArrayMsg:

    def test_roundtrip_with_dynamic_count(self):
        msg = CountFromArrayMsg()
        msg.num_items = 2
        for i in range(2):
            p = Point()
            p.x = (i + 1) * 5
            p.y = (i + 1) * 10
            msg.items.append(p)

        data = msg.encode_bytes()
        msg2 = CountFromArrayMsg.decode_bytes(data)
        assert msg2.num_items == 2
        assert len(msg2.items) == 2
        assert msg2.items[0].x == 5
        assert msg2.items[0].y == 10
        assert msg2.items[1].x == 10
        assert msg2.items[1].y == 20

    def test_zero_items(self):
        msg = CountFromArrayMsg()
        msg.num_items = 0
        data = msg.encode_bytes()
        msg2 = CountFromArrayMsg.decode_bytes(data)
        assert msg2.num_items == 0
        assert len(msg2.items) == 0

    def test_single_item(self):
        msg = CountFromArrayMsg()
        msg.num_items = 1
        p = Point()
        p.x = 42
        p.y = 84
        msg.items.append(p)
        data = msg.encode_bytes()
        msg2 = CountFromArrayMsg.decode_bytes(data)
        assert msg2.num_items == 1
        assert len(msg2.items) == 1
        assert msg2.items[0].x == 42
        assert msg2.items[0].y == 84

    def test_wire_size(self):
        """1 byte count + N * 4 bytes per Point."""
        msg = CountFromArrayMsg()
        msg.num_items = 3
        for _ in range(3):
            msg.items.append(Point())
        data = msg.encode_bytes()
        assert len(data) == 1 + 3 * 4

    def test_idempotent(self):
        msg = CountFromArrayMsg()
        msg.num_items = 2
        for i in range(2):
            p = Point()
            p.x = i
            p.y = i + 100
            msg.items.append(p)
        data1 = msg.encode_bytes()
        msg2 = CountFromArrayMsg.decode_bytes(data1)
        data2 = msg2.encode_bytes()
        assert data1 == data2


# ---------------------------------------------------------------------------
# ChoiceMsg with TypeABody
# ---------------------------------------------------------------------------
class TestChoiceMsgTypeA:

    def test_roundtrip(self):
        msg = ChoiceMsg()
        msg.msg_type = Constants.TYPE_A
        msg.length = 5
        body = TypeABody()
        body.sub_type = Constants.SUB_X
        body.sub_body = SubX()
        body.sub_body.val = 0x12345678
        msg.body = body

        data = msg.encode_bytes()
        msg2 = ChoiceMsg.decode_bytes(data)
        assert msg2.msg_type == Constants.TYPE_A
        assert isinstance(msg2.body, TypeABody)
        assert msg2.body.sub_type == Constants.SUB_X
        assert isinstance(msg2.body.sub_body, SubX)
        assert msg2.body.sub_body.val == 0x12345678

    def test_with_sub_y(self):
        msg = ChoiceMsg()
        msg.msg_type = Constants.TYPE_A
        msg.length = 5
        body = TypeABody()
        body.sub_type = Constants.SUB_Y
        body.sub_body = SubY()
        body.sub_body.a = 100
        body.sub_body.b = 200
        msg.body = body

        data = msg.encode_bytes()
        msg2 = ChoiceMsg.decode_bytes(data)
        assert isinstance(msg2.body, TypeABody)
        assert msg2.body.sub_type == Constants.SUB_Y
        assert isinstance(msg2.body.sub_body, SubY)
        assert msg2.body.sub_body.a == 100
        assert msg2.body.sub_body.b == 200


# ---------------------------------------------------------------------------
# ChoiceMsg with TypeBBody
# ---------------------------------------------------------------------------
class TestChoiceMsgTypeB:

    def test_roundtrip(self):
        msg = ChoiceMsg()
        msg.msg_type = Constants.TYPE_B
        msg.length = 4
        body = TypeBBody()
        body.tag = 0xCAFEBABE
        msg.body = body

        data = msg.encode_bytes()
        msg2 = ChoiceMsg.decode_bytes(data)
        assert msg2.msg_type == Constants.TYPE_B
        assert isinstance(msg2.body, TypeBBody)
        assert msg2.body.tag == 0xCAFEBABE

    def test_tag_zero(self):
        msg = ChoiceMsg()
        msg.msg_type = Constants.TYPE_B
        msg.length = 4
        body = TypeBBody()
        body.tag = 0
        msg.body = body

        data = msg.encode_bytes()
        msg2 = ChoiceMsg.decode_bytes(data)
        assert msg2.body.tag == 0

    def test_tag_max(self):
        msg = ChoiceMsg()
        msg.msg_type = Constants.TYPE_B
        msg.length = 4
        body = TypeBBody()
        body.tag = 0xFFFFFFFF
        msg.body = body

        data = msg.encode_bytes()
        msg2 = ChoiceMsg.decode_bytes(data)
        assert msg2.body.tag == 0xFFFFFFFF


# ---------------------------------------------------------------------------
# ChoiceMsg with FallbackBody
# ---------------------------------------------------------------------------
class TestChoiceMsgFallback:

    def test_roundtrip_unknown_type(self):
        msg = ChoiceMsg()
        msg.msg_type = 99  # not TYPE_A or TYPE_B
        msg.length = 4
        body = FallbackBody()
        body.raw = 0xDEAD1234
        msg.body = body

        data = msg.encode_bytes()
        msg2 = ChoiceMsg.decode_bytes(data)
        assert msg2.msg_type == 99
        assert isinstance(msg2.body, FallbackBody)
        assert msg2.body.raw == 0xDEAD1234

    def test_fallback_zero(self):
        msg = ChoiceMsg()
        msg.msg_type = 255
        msg.length = 4
        body = FallbackBody()
        body.raw = 0
        msg.body = body

        data = msg.encode_bytes()
        msg2 = ChoiceMsg.decode_bytes(data)
        assert isinstance(msg2.body, FallbackBody)
        assert msg2.body.raw == 0

    def test_multiple_unknown_types(self):
        """Different unknown type codes should all fall through to FallbackBody."""
        for type_code in [0, 3, 50, 128, 255]:
            msg = ChoiceMsg()
            msg.msg_type = type_code
            msg.length = 4
            body = FallbackBody()
            body.raw = type_code
            msg.body = body

            data = msg.encode_bytes()
            msg2 = ChoiceMsg.decode_bytes(data)
            assert isinstance(msg2.body, FallbackBody)
            assert msg2.body.raw == type_code


# ---------------------------------------------------------------------------
# TypeABody with SubX choice
# ---------------------------------------------------------------------------
class TestTypeABodySubX:

    def test_roundtrip(self):
        body = TypeABody()
        body.sub_type = Constants.SUB_X
        body.sub_body = SubX()
        body.sub_body.val = 0xDEADBEEF

        data = body.encode_bytes()
        body2 = TypeABody.decode_bytes(data)
        assert body2.sub_type == Constants.SUB_X
        assert isinstance(body2.sub_body, SubX)
        assert body2.sub_body.val == 0xDEADBEEF

    def test_sub_x_zero_val(self):
        body = TypeABody()
        body.sub_type = Constants.SUB_X
        body.sub_body = SubX()
        body.sub_body.val = 0

        data = body.encode_bytes()
        body2 = TypeABody.decode_bytes(data)
        assert body2.sub_body.val == 0

    def test_sub_x_max_val(self):
        body = TypeABody()
        body.sub_type = Constants.SUB_X
        body.sub_body = SubX()
        body.sub_body.val = 0xFFFFFFFF

        data = body.encode_bytes()
        body2 = TypeABody.decode_bytes(data)
        assert body2.sub_body.val == 0xFFFFFFFF


# ---------------------------------------------------------------------------
# TypeABody with SubY choice
# ---------------------------------------------------------------------------
class TestTypeABodySubY:

    def test_roundtrip(self):
        body = TypeABody()
        body.sub_type = Constants.SUB_Y
        body.sub_body = SubY()
        body.sub_body.a = 0x1111
        body.sub_body.b = 0x2222

        data = body.encode_bytes()
        body2 = TypeABody.decode_bytes(data)
        assert body2.sub_type == Constants.SUB_Y
        assert isinstance(body2.sub_body, SubY)
        assert body2.sub_body.a == 0x1111
        assert body2.sub_body.b == 0x2222

    def test_sub_y_max(self):
        body = TypeABody()
        body.sub_type = Constants.SUB_Y
        body.sub_body = SubY()
        body.sub_body.a = 0xFFFF
        body.sub_body.b = 0xFFFF

        data = body.encode_bytes()
        body2 = TypeABody.decode_bytes(data)
        assert body2.sub_body.a == 0xFFFF
        assert body2.sub_body.b == 0xFFFF


# ---------------------------------------------------------------------------
# SubX standalone
# ---------------------------------------------------------------------------
class TestSubXStruct:

    def test_default(self):
        s = SubX()
        assert s.val == 0

    def test_roundtrip(self):
        s = SubX()
        s.val = 42
        data = s.encode_bytes()
        s2 = SubX.decode_bytes(data)
        assert s2.val == 42

    def test_wire_size(self):
        s = SubX()
        s.val = 0
        assert len(s.encode_bytes()) == 4


# ---------------------------------------------------------------------------
# SubY standalone
# ---------------------------------------------------------------------------
class TestSubYStruct:

    def test_default(self):
        s = SubY()
        assert s.a == 0
        assert s.b == 0

    def test_roundtrip(self):
        s = SubY()
        s.a = 300
        s.b = 400
        data = s.encode_bytes()
        s2 = SubY.decode_bytes(data)
        assert s2.a == 300
        assert s2.b == 400

    def test_wire_size(self):
        s = SubY()
        assert len(s.encode_bytes()) == 4


# ---------------------------------------------------------------------------
# TypeBBody standalone
# ---------------------------------------------------------------------------
class TestTypeBBodyStruct:

    def test_default(self):
        b = TypeBBody()
        assert b.tag == 0

    def test_roundtrip(self):
        b = TypeBBody()
        b.tag = 0xABCDEF01
        data = b.encode_bytes()
        b2 = TypeBBody.decode_bytes(data)
        assert b2.tag == 0xABCDEF01


# ---------------------------------------------------------------------------
# FallbackBody standalone
# ---------------------------------------------------------------------------
class TestFallbackBodyStruct:

    def test_default(self):
        f = FallbackBody()
        assert f.raw == 0

    def test_roundtrip(self):
        f = FallbackBody()
        f.raw = 0x99887766
        data = f.encode_bytes()
        f2 = FallbackBody.decode_bytes(data)
        assert f2.raw == 0x99887766


# ---------------------------------------------------------------------------
# ChoiceMsg idempotent
# ---------------------------------------------------------------------------
class TestChoiceMsgIdempotent:

    @pytest.mark.parametrize("type_code,body_factory,body_len", [
        (1, lambda: _make_type_a_body(Constants.SUB_X), 5),  # sub_type(1) + SubX(4)
        (1, lambda: _make_type_a_body(Constants.SUB_Y), 5),  # sub_type(1) + SubY(4)
        (2, lambda: _make_type_b_body(), 4),                  # TypeBBody(4)
        (99, lambda: _make_fallback_body(), 4),               # FallbackBody(4)
    ])
    def test_encode_decode_encode_stable(self, type_code, body_factory, body_len):
        msg = ChoiceMsg()
        msg.msg_type = type_code
        msg.length = body_len
        msg.body = body_factory()

        data1 = msg.encode_bytes()
        msg2 = ChoiceMsg.decode_bytes(data1)
        data2 = msg2.encode_bytes()
        assert data1 == data2


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _make_type_a_body(sub_type):
    body = TypeABody()
    body.sub_type = sub_type
    if sub_type == Constants.SUB_X:
        body.sub_body = SubX()
        body.sub_body.val = 0x1234
    elif sub_type == Constants.SUB_Y:
        body.sub_body = SubY()
        body.sub_body.a = 10
        body.sub_body.b = 20
    return body


def _make_type_b_body():
    body = TypeBBody()
    body.tag = 0xABCD
    return body


def _make_fallback_body():
    body = FallbackBody()
    body.raw = 0xFFFF
    return body

File: conduit/tests/java/TestArraysChoices.java
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import static org.junit.jupiter.api.Assertions.*;

import java.util.ArrayList;
import java.util.List;

/**
 * Tests for array messages, choice/discriminator messages,
 * nested choices, and struct roundtrips.
 */
public class TestArraysChoices {

    // ========================================================================
    // Point struct tests
    // ========================================================================

    @Test
    @DisplayName("Point: default values roundtrip")
    void pointDefault() {
        arrays_choices.Point p = new arrays_choices.Point();
        byte[] encoded = p.encodeBytes();
        arrays_choices.Point decoded = arrays_choices.Point.decodeBytes(encoded);
        assertEquals(0, decoded.x);
        assertEquals(0, decoded.y);
    }

    @Test
    @DisplayName("Point: set values and roundtrip")
    void pointWithValues() {
        arrays_choices.Point p = new arrays_choices.Point();
        p.x = 100;
        p.y = 200;
        byte[] encoded = p.encodeBytes();
        arrays_choices.Point decoded = arrays_choices.Point.decodeBytes(encoded);
        assertEquals(100, decoded.x);
        assertEquals(200, decoded.y);
    }

    @Test
    @DisplayName("Point: max u16 values roundtrip")
    void pointMaxValues() {
        arrays_choices.Point p = new arrays_choices.Point();
        p.x = 65535;
        p.y = 65535;
        byte[] encoded = p.encodeBytes();
        arrays_choices.Point decoded = arrays_choices.Point.decodeBytes(encoded);
        assertEquals(65535, decoded.x);
        assertEquals(65535, decoded.y);
    }

    @Test
    @DisplayName("Point: encodeBytes produces 4 bytes")
    void pointEncodedSize() {
        arrays_choices.Point p = new arrays_choices.Point();
        assertEquals(4, p.encodeBytes().length);
    }

    // ========================================================================
    // FixedArrayMsg tests
    // ========================================================================

    @Test
    @DisplayName("FixedArrayMsg: roundtrip with 3 Points")
    void fixedArrayMsgRoundtrip() {
        arrays_choices.FixedArrayMsg msg = new arrays_choices.FixedArrayMsg();
        for (int i = 0; i < 3; i++) {
            arrays_choices.Point p = new arrays_choices.Point();
            p.x = (i + 1) * 10;
            p.y = (i + 1) * 20;
            msg.points.add(p);
        }
        byte[] encoded = msg.encodeBytes();
        arrays_choices.FixedArrayMsg decoded = arrays_choices.FixedArrayMsg.decodeBytes(encoded);
        assertEquals(3, decoded.points.size());
        assertEquals(10, decoded.points.get(0).x);
        assertEquals(20, decoded.points.get(0).y);
        assertEquals(20, decoded.points.get(1).x);
        assertEquals(40, decoded.points.get(1).y);
        assertEquals(30, decoded.points.get(2).x);
        assertEquals(60, decoded.points.get(2).y);
    }

    @Test
    @DisplayName("FixedArrayMsg: encoded size is 12 bytes (3 x 4)")
    void fixedArrayMsgEncodedSize() {
        arrays_choices.FixedArrayMsg msg = new arrays_choices.FixedArrayMsg();
        for (int i = 0; i < 3; i++) {
            msg.points.add(new arrays_choices.Point());
        }
        byte[] encoded = msg.encodeBytes();
        assertEquals(12, encoded.length);  // 3 Points * 4 bytes each
    }

    @Test
    @DisplayName("FixedArrayMsg: all zero points roundtrip")
    void fixedArrayMsgZeroPoints() {
        arrays_choices.FixedArrayMsg msg = new arrays_choices.FixedArrayMsg();
        for (int i = 0; i < 3; i++) {
            msg.points.add(new arrays_choices.Point());
        }
        byte[] encoded = msg.encodeBytes();
        arrays_choices.FixedArrayMsg decoded = arrays_choices.FixedArrayMsg.decodeBytes(encoded);
        assertEquals(3, decoded.points.size());
        for (arrays_choices.Point p : decoded.points) {
            assertEquals(0, p.x);
            assertEquals(0, p.y);
        }
    }

    @Test
    @DisplayName("FixedArrayMsg: max value points roundtrip")
    void fixedArrayMsgMaxPoints() {
        arrays_choices.FixedArrayMsg msg = new arrays_choices.FixedArrayMsg();
        for (int i = 0; i < 3; i++) {
            arrays_choices.Point p = new arrays_choices.Point();
            p.x = 65535;
            p.y = 65535;
            msg.points.add(p);
        }
        byte[] encoded = msg.encodeBytes();
        arrays_choices.FixedArrayMsg decoded = arrays_choices.FixedArrayMsg.decodeBytes(encoded);
        for (arrays_choices.Point p : decoded.points) {
            assertEquals(65535, p.x);
            assertEquals(65535, p.y);
        }
    }

    // ========================================================================
    // CountFromArrayMsg tests
    // ========================================================================

    @Test
    @DisplayName("CountFromArrayMsg: roundtrip with 2 items")
    void countFromArrayMsg2Items() {
        arrays_choices.CountFromArrayMsg msg = new arrays_choices.CountFromArrayMsg();
        msg.numItems = 2;
        for (int i = 0; i < 2; i++) {
            arrays_choices.Point p = new arrays_choices.Point();
            p.x = (i + 1) * 100;
            p.y = (i + 1) * 200;
            msg.items.add(p);
        }
        byte[] encoded = msg.encodeBytes();
        arrays_choices.CountFromArrayMsg decoded = arrays_choices.CountFromArrayMsg.decodeBytes(encoded);
        assertEquals(2, decoded.numItems);
        assertEquals(2, decoded.items.size());
        assertEquals(100, decoded.items.get(0).x);
        assertEquals(200, decoded.items.get(0).y);
        assertEquals(200, decoded.items.get(1).x);
        assertEquals(400, decoded.items.get(1).y);
    }

    @Test
    @DisplayName("CountFromArrayMsg: roundtrip with 0 items")
    void countFromArrayMsg0Items() {
        arrays_choices.CountFromArrayMsg msg = new arrays_choices.CountFromArrayMsg();
        msg.numItems = 0;
        byte[] encoded = msg.encodeBytes();
        arrays_choices.CountFromArrayMsg decoded = arrays_choices.CountFromArrayMsg.decodeBytes(encoded);
        assertEquals(0, decoded.numItems);
        assertEquals(0, decoded.items.size());
    }

    @Test
    @DisplayName("CountFromArrayMsg: roundtrip with 5 items")
    void countFromArrayMsg5Items() {
        arrays_choices.CountFromArrayMsg msg = new arrays_choices.CountFromArrayMsg();
        msg.numItems = 5;
        for (int i = 0; i < 5; i++) {
            arrays_choices.Point p = new arrays_choices.Point();
            p.x = i;
            p.y = i * 10;
            msg.items.add(p);
        }
        byte[] encoded = msg.encodeBytes();
        arrays_choices.CountFromArrayMsg decoded = arrays_choices.CountFromArrayMsg.decodeBytes(encoded);
        assertEquals(5, decoded.numItems);
        assertEquals(5, decoded.items.size());
        for (int i = 0; i < 5; i++) {
            assertEquals(i, decoded.items.get(i).x);
            assertEquals(i * 10, decoded.items.get(i).y);
        }
    }

    @Test
    @DisplayName("CountFromArrayMsg: encoded size is 1 + numItems*4")
    void countFromArrayMsgEncodedSize() {
        arrays_choices.CountFromArrayMsg msg = new arrays_choices.CountFromArrayMsg();
        msg.numItems = 3;
        for (int i = 0; i < 3; i++) {
            msg.items.add(new arrays_choices.Point());
        }
        byte[] encoded = msg.encodeBytes();
        assertEquals(1 + 3 * 4, encoded.length); // 1 byte count + 3*4 bytes points
    }

    // ========================================================================
    // ChoiceMsg tests
    // ========================================================================

    @Test
    @DisplayName("ChoiceMsg: TypeABody choice roundtrip")
    void choiceMsgTypeA() {
        arrays_choices.ChoiceMsg msg = new arrays_choices.ChoiceMsg();
        msg.msgType = (int) arrays_choices.Constants.TYPE_A;
        msg.length = 5; // TypeABody: subType(1) + SubX.val(4)
        arrays_choices.TypeABody typeA = new arrays_choices.TypeABody();
        typeA.subType = (int) arrays_choices.Constants.SUB_X;
        arrays_choices.SubX subX = new arrays_choices.SubX();
        subX.val = 12345;
        typeA.subBody = subX;
        msg.body = typeA;

        byte[] encoded = msg.encodeBytes();
        arrays_choices.ChoiceMsg decoded = arrays_choices.ChoiceMsg.decodeBytes(encoded);
        assertEquals((int) arrays_choices.Constants.TYPE_A, decoded.msgType);
        assertInstanceOf(arrays_choices.TypeABody.class, decoded.body);
        arrays_choices.TypeABody decodedTypeA = (arrays_choices.TypeABody) decoded.body;
        assertEquals((int) arrays_choices.Constants.SUB_X, decodedTypeA.subType);
        assertInstanceOf(arrays_choices.SubX.class, decodedTypeA.subBody);
        assertEquals(12345, ((arrays_choices.SubX) decodedTypeA.subBody).val);
    }

    @Test
    @DisplayName("ChoiceMsg: TypeBBody choice roundtrip")
    void choiceMsgTypeB() {
        arrays_choices.ChoiceMsg msg = new arrays_choices.ChoiceMsg();
        msg.msgType = (int) arrays_choices.Constants.TYPE_B;
        msg.length = 4; // TypeBBody: tag(4)
        arrays_choices.TypeBBody typeB = new arrays_choices.TypeBBody();
        typeB.tag = 0xABCDEF;
        msg.body = typeB;

        byte[] encoded = msg.encodeBytes();
        arrays_choices.ChoiceMsg decoded = arrays_choices.ChoiceMsg.decodeBytes(encoded);
        assertEquals((int) arrays_choices.Constants.TYPE_B, decoded.msgType);
        assertInstanceOf(arrays_choices.TypeBBody.class, decoded.body);
        assertEquals(0xABCDEF, ((arrays_choices.TypeBBody) decoded.body).tag);
    }

    @Test
    @DisplayName("ChoiceMsg: Fallback choice for unknown msgType")
    void choiceMsgFallback() {
        // Build the raw bytes manually for an unknown msgType
        arrays_choices.codec.BitWriter w = new arrays_choices.codec.BitWriter();
        w.writeU8(99);          // unknown msgType
        w.writeU16(4, true);    // length = 4 bytes for FallbackBody
        w.writeU32(0xDEADBEEF, true); // raw data for FallbackBody
        byte[] data = w.toBytes();

        arrays_choices.ChoiceMsg decoded = arrays_choices.ChoiceMsg.decodeBytes(data);
        assertEquals(99, decoded.msgType);
        assertInstanceOf(arrays_choices.FallbackBody.class, decoded.body);
        assertEquals(0xDEADBEEF, ((arrays_choices.FallbackBody) decoded.body).raw);
    }

    // ========================================================================
    // TypeABody nested choice tests
    // ========================================================================

    @Test
    @DisplayName("TypeABody: SubX nested choice roundtrip")
    void typeABodySubX() {
        arrays_choices.TypeABody msg = new arrays_choices.TypeABody();
        msg.subType = (int) arrays_choices.Constants.SUB_X;
        arrays_choices.SubX sub = new arrays_choices.SubX();
        sub.val = 999;
        msg.subBody = sub;

        byte[] encoded = msg.encodeBytes();
        arrays_choices.TypeABody decoded = arrays_choices.TypeABody.decodeBytes(encoded);
        assertEquals((int) arrays_choices.Constants.SUB_X, decoded.subType);
        assertInstanceOf(arrays_choices.SubX.class, decoded.subBody);
        assertEquals(999, ((arrays_choices.SubX) decoded.subBody).val);
    }

    @Test
    @DisplayName("TypeABody: SubY nested choice roundtrip")
    void typeABodySubY() {
        arrays_choices.TypeABody msg = new arrays_choices.TypeABody();
        msg.subType = (int) arrays_choices.Constants.SUB_Y;
        arrays_choices.SubY sub = new arrays_choices.SubY();
        sub.a = 100;
        sub.b = 200;
        msg.subBody = sub;

        byte[] encoded = msg.encodeBytes();
        arrays_choices.TypeABody decoded = arrays_choices.TypeABody.decodeBytes(encoded);
        assertEquals((int) arrays_choices.Constants.SUB_Y, decoded.subType);
        assertInstanceOf(arrays_choices.SubY.class, decoded.subBody);
        arrays_choices.SubY decodedSub = (arrays_choices.SubY) decoded.subBody;
        assertEquals(100, decodedSub.a);
        assertEquals(200, decodedSub.b);
    }

    @Test
    @DisplayName("TypeABody: unknown subType throws exception")
    void typeABodyUnknownSubType() {
        arrays_choices.codec.BitWriter w = new arrays_choices.codec.BitWriter();
        w.writeU8(99);  // unknown subType
        byte[] data = w.toBytes();
        assertThrows(arrays_choices.codec.ConduitCodecException.class, () -> {
            arrays_choices.TypeABody.decodeBytes(data);
        });
    }

    // ========================================================================
    // TypeBBody tests
    // ========================================================================

    @Test
    @DisplayName("TypeBBody: roundtrip")
    void typeBBodyRoundtrip() {
        arrays_choices.TypeBBody msg = new arrays_choices.TypeBBody();
        msg.tag = 0x12345678;
        byte[] encoded = msg.encodeBytes();
        arrays_choices.TypeBBody decoded = arrays_choices.TypeBBody.decodeBytes(encoded);
        assertEquals(0x12345678, decoded.tag);
    }

    @Test
    @DisplayName("TypeBBody: zero tag roundtrip")
    void typeBBodyZero() {
        arrays_choices.TypeBBody msg = new arrays_choices.TypeBBody();
        msg.tag = 0;
        byte[] encoded = msg.encodeBytes();
        arrays_choices.TypeBBody decoded = arrays_choices.TypeBBody.decodeBytes(encoded);
        assertEquals(0, decoded.tag);
    }

    // ========================================================================
    // SubX / SubY standalone tests
    // ========================================================================

    @Test
    @DisplayName("SubX: roundtrip")
    void subXRoundtrip() {
        arrays_choices.SubX msg = new arrays_choices.SubX();
        msg.val = 0xFFFF;
        byte[] encoded = msg.encodeBytes();
        arrays_choices.SubX decoded = arrays_choices.SubX.decodeBytes(encoded);
        assertEquals(0xFFFF, decoded.val);
    }

    @Test
    @DisplayName("SubY: roundtrip")
    void subYRoundtrip() {
        arrays_choices.SubY msg = new arrays_choices.SubY();
        msg.a = 5000;
        msg.b = 6000;
        byte[] encoded = msg.encodeBytes();
        arrays_choices.SubY decoded = arrays_choices.SubY.decodeBytes(encoded);
        assertEquals(5000, decoded.a);
        assertEquals(6000, decoded.b);
    }

    // ========================================================================
    // FallbackBody tests
    // ========================================================================

    @Test
    @DisplayName("FallbackBody: roundtrip")
    void fallbackBodyRoundtrip() {
        arrays_choices.FallbackBody msg = new arrays_choices.FallbackBody();
        msg.raw = 0xCAFEBABE;
        byte[] encoded = msg.encodeBytes();
        arrays_choices.FallbackBody decoded = arrays_choices.FallbackBody.decodeBytes(encoded);
        assertEquals(0xCAFEBABE, decoded.raw);
    }

    // ========================================================================
    // Constants tests
    // ========================================================================

    @Test
    @DisplayName("Constants: TYPE_A is 1")
    void constantsTypeA() {
        assertEquals(1, arrays_choices.Constants.TYPE_A);
    }

    @Test
    @DisplayName("Constants: TYPE_B is 2")
    void constantsTypeB() {
        assertEquals(2, arrays_choices.Constants.TYPE_B);
    }

    @Test
    @DisplayName("Constants: SUB_X is 10")
    void constantsSubX() {
        assertEquals(10, arrays_choices.Constants.SUB_X);
    }

    @Test
    @DisplayName("Constants: SUB_Y is 20")
    void constantsSubY() {
        assertEquals(20, arrays_choices.Constants.SUB_Y);
    }

    // ========================================================================
    // Deep nested roundtrip: ChoiceMsg -> TypeABody -> SubX/SubY
    // ========================================================================

    @Test
    @DisplayName("Deep nested: ChoiceMsg -> TypeABody -> SubY roundtrip")
    void deepNestedChoiceSubY() {
        arrays_choices.ChoiceMsg msg = new arrays_choices.ChoiceMsg();
        msg.msgType = (int) arrays_choices.Constants.TYPE_A;
        msg.length = 5; // TypeABody: subType(1) + SubY(2+2)
        arrays_choices.TypeABody typeA = new arrays_choices.TypeABody();
        typeA.subType = (int) arrays_choices.Constants.SUB_Y;
        arrays_choices.SubY sub = new arrays_choices.SubY();
        sub.a = 1111;
        sub.b = 2222;
        typeA.subBody = sub;
        msg.body = typeA;

        byte[] encoded = msg.encodeBytes();
        arrays_choices.ChoiceMsg decoded = arrays_choices.ChoiceMsg.decodeBytes(encoded);

        assertInstanceOf(arrays_choices.TypeABody.class, decoded.body);
        arrays_choices.TypeABody decodedA = (arrays_choices.TypeABody) decoded.body;
        assertInstanceOf(arrays_choices.SubY.class, decodedA.subBody);
        arrays_choices.SubY decodedSub = (arrays_choices.SubY) decodedA.subBody;
        assertEquals(1111, decodedSub.a);
        assertEquals(2222, decodedSub.b);
    }

    @Test
    @DisplayName("Deep nested: ChoiceMsg -> TypeABody -> SubX roundtrip with max values")
    void deepNestedChoiceSubXMax() {
        arrays_choices.ChoiceMsg msg = new arrays_choices.ChoiceMsg();
        msg.msgType = (int) arrays_choices.Constants.TYPE_A;
        msg.length = 5; // TypeABody: subType(1) + SubX.val(4)
        arrays_choices.TypeABody typeA = new arrays_choices.TypeABody();
        typeA.subType = (int) arrays_choices.Constants.SUB_X;
        arrays_choices.SubX sub = new arrays_choices.SubX();
        sub.val = 0x7FFFFFFF;
        typeA.subBody = sub;
        msg.body = typeA;

        byte[] encoded = msg.encodeBytes();
        arrays_choices.ChoiceMsg decoded = arrays_choices.ChoiceMsg.decodeBytes(encoded);

        arrays_choices.TypeABody decodedA = (arrays_choices.TypeABody) decoded.body;
        arrays_choices.SubX decodedSub = (arrays_choices.SubX) decodedA.subBody;
        assertEquals(0x7FFFFFFF, decodedSub.val);
    }
}
